cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -240,6 +261,36 @@ Status StatSelector(const PlatformFilename& dir_fn, const FileSelector& select,
   return Status::OK();
 }
 
+Status IdentifyFileSelector(const PlatformFilename& dir_fn, const FileSelector& select,
```

Comment:
It seems to share a lot of common logic from `StatSelector`. Is is possible to modify it directly?

Reply:
`IdentifyFile` requires an `std::filesystem::path` while `StatFile` requires `std::wstring` or `std::string`. Because it can use `std::filesystem::path` `IdentifyFile` takes advantage of the performance boost of `directory_iterator` for its loop. Basically, There is enough unique logic that combining them is looking uglier. The main thing they have in common is the boilerplate at the start that checks for the existence of the file. I think the better thing to do is to factor out that code, instead of trying to combine the rest.